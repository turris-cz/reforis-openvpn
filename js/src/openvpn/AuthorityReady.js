/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useContext } from "react";
import PropTypes from "prop-types";

import { Button, useAPIDelete, AlertContext } from "foris";

import API_URLs from "API";

AuthorityReady.propTypes = {
    serverEnabled: PropTypes.bool.isRequired,
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityReady({ serverEnabled, onReload }) {
    const setAlert = useContext(AlertContext);

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.isSuccess) {
            onReload();
        } else if (deleteResponse.isError) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onReload, setAlert]);

    return (
        <>
            <h3>{_("Certificate authority")}</h3>
            {serverEnabled
                ? <p>{_("You can't delete the CA while the OpenVPN server is enabled. To delete the CA you need to disable the server configuration first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority (CA) is set up properly.")}</p>
                        <Button onClick={() => deleteCA()}>{_("Delete CA")}</Button>
                    </>
                )}
        </>
    );
}
