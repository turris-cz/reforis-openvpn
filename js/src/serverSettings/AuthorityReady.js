/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { Button, useAPIDelete, useAlert, API_STATE } from "foris";

import API_URLs from "API";

AuthorityReady.propTypes = {
    serverEnabled: PropTypes.bool.isRequired,
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityReady({ serverEnabled, onReload }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onReload();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onReload, setAlert]);

    return (
        <div className="card p-4">
            <h2>{_("Certificate authority")}</h2>
            {serverEnabled ? (
                <p>
                    {_(
                        "You can't delete the CA while the OpenVPN server is enabled. To delete the CA you need to disable the server configuration first."
                    )}
                </p>
            ) : (
                <p>
                    {_(
                        "Your certificate authority (CA) is set up properly. Please note that if you delete it all clients will have their access revoked."
                    )}
                </p>
            )}
            <div className="text-right">
                <Button
                    onClick={() => deleteCA()}
                    className="btn-danger"
                    forisFormSize
                    loading={deleteResponse.state === API_STATE.SENDING}
                    disabled={serverEnabled}
                >
                    {_("Delete CA")}
                </Button>
            </div>
        </div>
    );
}
