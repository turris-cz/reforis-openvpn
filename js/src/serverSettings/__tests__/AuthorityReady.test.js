/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, fireEvent, wait } from "foris/testUtils/customTestRender";
import mockAxios from 'jest-mock-axios';
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";

import AuthorityReady from "../AuthorityReady";


describe("<AuthorityReady />", () => {
    let componentContainer;
    const handleReload = jest.fn();

    beforeEach(() => {
        const { container } = render(<AuthorityReady onReload={handleReload} serverEnabled={false} />);
        componentContainer = container;
    });

    it("should match snapshot", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should send request when button is clicked", () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        expect(mockAxios.delete).toBeCalledWith("/reforis/openvpn/api/authority", expect.anything());
    });

    it("should handle error", async () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith("Cannot delete certificate authority");
        });
    });

    it("should handle success", async () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        mockAxios.mockResponse({});
        await wait(() => {
            expect(handleReload).toHaveBeenCalledTimes(1);
        });
    });

    it("should prevent deletion when server is enabled", () => {
        const { container } = render(<AuthorityReady onReload={handleReload} serverEnabled={true} />);
        expect(container).toMatchSnapshot();
    });
});
