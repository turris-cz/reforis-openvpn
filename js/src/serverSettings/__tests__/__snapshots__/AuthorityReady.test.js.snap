// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityReady /> should match snapshot 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Your certificate authority (CA) is set up properly. Please note that if you delete it, all clients will have their access revoked.
    </p>
    <div
      class="text-right"
    >
      <button
        class="btn btn-danger col-sm-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Delete CA
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<AuthorityReady /> should prevent deletion when server is enabled 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      You can't delete the CA while the OpenVPN server is enabled. To delete the CA, you need to disable the server configuration first.
    </p>
    <div
      class="text-right"
    >
      <button
        class="btn btn-danger col-sm-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        disabled=""
        type="button"
      >
        <span>
          Delete CA
        </span>
      </button>
    </div>
  </div>
</div>
`;
